File: KAKAO_BLIND_RECRUITMENT/2019/후보키/solve.cpp
#include <string>
#include <vector>
#include <iostream>
#include <set>

using namespace std;

int solution(vector<vector<string>> relation) {
    int answer = 0;
    
    int row = relation.size();
    int col = relation[0].size();
    
    vector<int> bit_storage;
    
    // column 정보에 대한 조합을 bitmasking으로 만듬
    for(int i=1; i<(1<<col); i++){
        set<string> sets;
        
        for(int j=0; j<row; j++){
            string str = "";
            
            // 비트가 1인 경우 str에 더해서 문자열을 만듬
            for(int k=0; k<col; k++){
                if(i&(1<<k)) str += relation[j][k];
            }
            sets.insert(str);
        }
        // set의 크기가 row와 일치하지 않는다면 중복 문자열이 존재한다는 의미
        if(sets.size() == row){
            bool flag = true;
            // 중복 문자열이 존재하지 않더라도 최소성을 만족하기 위해서는 비트 연산의 결과가 1이 나오면 안됨
            for(auto bit : bit_storage){
                // 그냥 i & bit로만 계산하였을 경우에는 1001(9) 와 0101(5)와 같이 실제로 부분집합에 포함되지 않는 경우가 존재할 수 있음
                if((i & bit) == bit){
                    flag = false;
                    break;
                }
            }
            if(flag) bit_storage.push_back(i);
        }
    }
    
    answer = bit_storage.size();
    
    return answer;
}

File: 월간_코드_챌린지/시즌1/삼각_달팽이/solve.py
def solution(n):
    answer = []
    
    triangle = [[0]*x for x in range(1,n+1)]
    
    row = -1
    col = 0
    number = 1
    
    # n번 방향을 바꿈 왼쪽 대각선 -> 오른쪽 -> 오른쪽 대각선
    for i in range(n):
        for j in range(i,n):
            # 왼쪽 대각선
            if i%3 == 0:
                row += 1
            # 오른쪽 
            elif i%3 == 1:
                col += 1
            # 오른쪽 대각선
            else:
                row -= 1
                col -= 1
            triangle[row][col] = number
            number += 1
    
    answer = []
    
    for tri in triangle:
        for angle in tri:
            answer.append(angle)
            
    return answer
